python-package/lightgbm/basic.py

```diff
@@ -2762,18 +2765,7 @@ def set_label(self, label: Optional[_LGBM_LabelType]) -> "Dataset":
             if isinstance(label, pd_DataFrame):
                 if len(label.columns) > 1:
                     raise ValueError('DataFrame for label cannot have multiple columns')
-                _check_for_bad_pandas_dtypes(label.dtypes)
-                try:
-                    # most common case (no nullable dtypes)
-                    label = label.to_numpy(dtype=np.float32, copy=False)
-                except TypeError:
-                    # 1.0 <= pd version < 1.1 and nullable dtypes, least common case
-                    # raises error because array is casted to type(pd.NA) and there's no na_value argument
-                    label = label.astype(np.float32, copy=False).values
-                except ValueError:
-                    # data has nullable dtypes, but we can specify na_value argument and copy will be made
-                    label = label.to_numpy(dtype=np.float32, na_value=np.nan)
-                label_array = np.ravel(label)
+                label_array = np.ravel(_pandas_to_numpy(label))
```

Comment:
The label was always cast to float32, regardless of its input type. I think we should preserve that, so maybe the `_pandas_to_numpy` function could take the target data type or the list of data types and resolve it there. WDYT?

Reply:
ooooo thank you for pointing that out! Sorry I missed that.

I like the idea of passing in the list of target dtypes. That'll make it clearer that `label` always gets cast as `np.float32`. I'll add that, thank you!

Reply:
Updated this in https://github.com/microsoft/LightGBM/pull/6156/commits/4e84a52603a1d8527278aa09355bdf76f5360b9f.